File: leetcode/Algorithms/144.BinaryTreePreorderTraversal/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        self.helper(result, root)
        return result

    def helper(self, result, node):
        if node:
            result.append(node.val)
            self.helper(result, node.left)
            self.helper(result, node.right)

File: leetcode/Algorithms/144.BinaryTreePreorderTraversal/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();

        while (root != null || !stack.isEmpty()) {
            if (root != null) {
                result.add(root.val);
                if (root.right != null) stack.push(root.right);
                root = root.left;
            } else {
                root = stack.pop();
            }
        }

        return result;
    }
}
